sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+
+use std::fs::File;
+use std::io::Write;
+use std::process::{Command, Stdio, Child};
+use crate::io::jtag::{Jtag, JtagParams};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use std::cell::Cell;
+
+use std::net::TcpStream;
+
+use crate::impl_serializable_error;
+
+use std::{thread, time};
+
+
+/// Represents an OpenOCD server and we can interact with interactively
+pub struct OpenOcdClient {
+    /// OpenOCD child process
+    child: Child,
+    /// TcpStream to the telnet port of OpenOCD
+    telnet_stream: Cell<Option<TcpStream>>,
+}
+
+/// Errors related to the OpenOCD client.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdClientError {
+    #[error("No OpenOCD binary path provided")]
+    OpenOcdBinaryRequired,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(OpenOcdClientError);
+
+impl OpenOcdClient {
+    pub fn new(opts: &JtagParams) -> Result<OpenOcdClient> {
+        let mut args = Vec::new();
+        if let Some(cfg) = &opts.openocd_adapter_config {
+            args.push("-f");
+            args.push(&cfg);
+        }
+        // Tell OpenOCD how to connect to the lifecycle controller's RISC-V TAP.
+        // This works because lc_ctrl uses an instance of the RISC-V debug
+        // module [0]. Its IRLEN is hardcoded to 5 [1] and its IDCODE is 1 [2].
+        //
+        // [0]: https://docs.opentitan.org/hw/ip/lc_ctrl/doc/#life-cycle-tap-controller
+        // [1]: See where `dmi_jtag_tap` is created in hw/vendor/pulp_riscv_dbg/src/dmi_jtag.sv
+        // [2]: See where `dmi_jtag` is instantiated in hw/ip/lc_ctrl/rtl/lc_ctrl.sv
+        //
+        // Because these commands are command-line args, they *must* end with
+        // semicolons to satisfy the TCL lexer. When a TCL program is read from
+        // a file, newlines are sufficient to separate statements.
+        args.extend(vec![
+            "-c",
+            "transport select jtag;",
+            "-c",
+            "jtag newtap lc_ctrl tap -irlen 5 -expected-id 0x00000001 -ignore-bypass;",
+            "-c",
+            "target create lc_ctrl.tap.0 riscv -chain-position lc_ctrl.tap -rtos hwthread;",
```

Comment:
Maybe we should move these to a [target config file](https://openocd.org/doc/html/Config-File-Guidelines.html#Target-Config-Files), like we have for the CPU TAP: https://cs.opensource.google/opentitan/opentitan/+/master:util/openocd/target/lowrisc-earlgrey.cfg?q=newtap&ss=opentitan%2Fopentitan ? Then make the target config file a data dep of the build rule for otlib.

If so, then we probably want to make the existing CPU TAP target config file a data dep too, and decide which one to load based on the TAP strap pin state?

It seems our OT chip muxes which TAP is routed to the JTAG pins, rather than daisy chaining the TAPs internally, so I'm guessing we can't just declare multiple TAPs in the target config file?

If that is **_not_** true, then could we declare both taps in same target config file and select them dynamically after the server starts?

Reply:
Yes exactly, because of the multiplexing we can't declare both in the same file, we'll probably need to restart OpenOCD when we change TAPs. I think moving those to a config files like for the cpu is the way to go, I'll do that when the sequence is finalized. I was thinking we might want the jtag transport function to take a parameter specifying the TAP to use and it will change strap pins and load the corresponding openocd file? (and maybe reset the chip if the TAP can't be change dynamically in this life cycle)

Reply:
Do we want the `jtag()` function in `.../transport/cw310/mod.rs` to take this added parameter of what TAP to use? or do we want to pass this added parameter to the `connect()` function that you mentioned below will be what spawns the OpenOCD server? I would think the latter no?